learn/appendix_opengl/KHR_debug.md

```diff
@@ -0,0 +1,194 @@
+);
+```
+The parameter `source` is the message source filter. It can be your own
+application or components of the OpenGL or windowing API. `type` filters which
+message types to control. `severity` filters which severity to control. In
+addition to this, the enum value `.DontCare` does not filter any message type.
+When `numIds` is positive, ids is an array of message IDs to filter. You can use
+`0` and `System.Array.Empty<int>()` if you don't want to feed the function any
+message IDs. And finally `isEnabled` controls if the filtered messages are to be
+permitted or ignored.
+
+This function is especially useful for turning down warnings that you do not
+desire to see. Since the messages generated are driver specific, it might be
+useful to keep a list of IDs you might want to mute or permit per vendor, that
+way your debugging logs are clear of any frequent informational messages.
+
+### `GL.DebugMessageInsert()`
+The following function allows you to insert a custom message into the debug
+message stream. The parameters are similar to that of `GL.DebugMessageControl`
+filter parameters.
+```cs
+void GL.DebugMessageInsert(
+    DebugSourceExternal source, /* Source of the message. */
+    DebugType type,             /* Type of the message. */
+    int id,                     /* An ID to associate with this message. */
+    DebugSeverity severity,     /* Severity of the message. */
+    int length,                 /* Length of the message or -1 for automatic. */
+    string message              /* The message to insert. */
+);
+```
+
+### `GL.GetDebugMessageLog()`
+In the abscense of a debug message callback, this function can be used to
+retreive debug messages that have been enqueued by the OpenGL context.
+```cs
+int GL.GetDebugMessageLog(
+    int number,                 /* Number of messages to retreive. */
+    int bufferSize,             /* Message buffer size. */
+    out DebugSource sources,    /* Retreived message sources. */
+    out DebugType types,        /* Retreived message types. */
+    out int ids,                /* Retreived message IDs. */
+    out DebugSeverity severity, /* Retreived message severities */
+    out int length,             /* Retreived message lengths. */
+    out string message          /* Retreived messages. */
+);  /* Returns: The number of messages retreived. */
+
+/* Note: an array overload also exists. */
+```
+
+### Additional Enables
+These additional enables are introduced to the API which allows you to control
+the debug messaging with `GL.Enable()` and `GL.Disable()`.
+
+| EnableCap | Description |
+|:----------|:------------|
+| `EnableCap.DebugOutput` | Enable or disable debug message generation. If disabled, inhibits sending messages to your callback. |
+| `EnableCap.DebugOutputSynchronous` | When enabled, debug messages are sent only to the thread that generated the message. Useful when debugging and viewing the callstack. Hinders performance. |
+
+Object Label API
```

Comment:
This and Debug Group APIs need to be higher up in the document as these are the functions that are actually interesting. 
Sending and receiving messages manually isn't as interesting and should come later in the article.